feat(gen2-migration): add dynamodb support for storage category

dgandhi62 opened the pull request:
## Add comprehensive DynamoDB support to Amplify Gen2 migration storage category

### Summary
Implemented full DynamoDB table migration support for Amplify Gen1 to Gen2 migration, enabling automatic conversion of DynamoDB resources from CloudFormation-based Gen1 projects to CDK-based Gen2 backend definitions.

### Key Features Added

**DynamoDB Table Discovery & Parsing:**
- Added `AppStorageDefinitionFetcher` to extract DynamoDB table configurations from Gen1 amplify-meta.json and cli-inputs.json
- Implemented AWS SDK integration to fetch live table schemas using `DescribeTableCommand`
- Support for parsing table attributes, key schemas, billing modes, and stream configurations

**Table Schema Migration:**
- **Partition & Sort Keys**: Automatic detection and conversion of primary key structures
- **Global Secondary Indexes (GSIs)**: Full GSI migration with partition/sort key mapping
- **Billing Modes**: Support for both `PAY_PER_REQUEST` and `PROVISIONED` with capacity units
- **DynamoDB Streams**: Stream configuration migration with view type preservation
- **Attribute Type Mapping**: Conversion from DynamoDB types (S, N, B) to CDK AttributeType enums

**Code Generation:**
- **CDK Table Creation**: Generates proper `new Table()` constructs with all configurations
- **Variable Name Sanitization**: Converts table names with hyphens to valid JavaScript identifiers
- **Import Management**: Automatic addition of required CDK imports (`Table`, `AttributeType`, `BillingMode`, `StreamViewType`)
- **Stack Integration**: Smart stack creation/reuse when S3 storage exists

**Comprehensive Test Coverage:**
- Unit tests for table parsing, GSI handling, Lambda permissions, and stream configurations
- Mock AWS SDK responses with proper TypeScript typing
- Edge case handling for multiple tables and mixed storage scenarios

### Technical Implementation

**Storage Definition Structure:**
```typescript
interface DynamoDBTableDefinition {
  tableName: string;
  partitionKey: { name: string; type: 'STRING' | 'NUMBER' | 'BINARY' };
  sortKey?: { name: string; type: 'STRING' | 'NUMBER' | 'BINARY' };
  gsis?: GlobalSecondaryIndex[];
  billingMode: 'PAY_PER_REQUEST' | 'PROVISIONED';
  readCapacity?: number;
  writeCapacity?: number;
  streamEnabled?: boolean;
  streamViewType?: string;
  lambdaPermissions?: LambdaPermission[];
  triggerFunctions?: string[];
}
```

Generated code example

```
const userTable_dev = new Table(storageStack, "userTable_dev", {
  partitionKey: { name: "userId", type: AttributeType.STRING },
  sortKey: { name: "timestamp", type: AttributeType.NUMBER },
  billingMode: BillingMode.PAY_PER_REQUEST,
  stream: StreamViewType.NEW_AND_OLD_IMAGES
});
```

The following still needs to be generated

**Lambda Integration:**
- **Environment Variables**: Automatic detection of Lambda functions accessing tables via env vars
- **IAM Permissions**: Generation of `grantReadWriteData` calls for Lambda table access
- **Stream Triggers**: Support for DynamoDB stream-triggered Lambda functions with proper permissions

```
userTable_dev.grantReadWriteData(backend.myFunction.resources.lambda);
backend.myFunction.addEnvironment("USER_TABLE_NAME", userTable_dev.tableName);
```

dgandhi62 committed:
feat: add dynamodb support for storage v2

dgandhi62 committed:
feat: add comprehensive DynamoDB migration support to Gen1→Gen2 tool

- Add DynamoDB table structure migration (partition key, sort key, GSIs)
- Fetch current AWS table configuration (billing mode, throughput, streams)
- Generate CDK Table constructs directly in backend.ts
- Add Lambda permissions and environment variables for table access
- Support all Gen1 indexable field types (string, number, binary)
- Use proper TypeScript interfaces instead of any types
- Fail fast on function configuration errors for data integrity

Missing: DynamoDB triggers (triggerFunctions) - requires EventSourceMapping

dgandhi62 committed:
feat: change fetcher to fetch all dyanmodb storage resource configs live

dgandhi62 committed:
fix: use underscores in dynamodb table names and transform handlers to es modules

- Replace hyphens with underscores in DynamoDB table name generation
- Update all table name references to use underscore format
- Transform CommonJS handler files to ES modules during migration
- Convert require statements to import statements
- Change exports.handler to export const handler with type annotations

dgandhi62 committed:
chore: fix test by adding bucketname since it is a required property

dgandhi62 committed:
chore: remove unused interface parameter in the storage generator

dgandhi62 committed:
chore: add back the dynamotables property

dgandhi62 committed:
chore: add comment

dgandhi62 committed:
feat: remove unnecessary comment

iliapolo commented on packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/adapters/storage/gen1_storage_codegen_adapter.ts:
```diff
@@ -1,15 +1,53 @@
+    functionName: string;
+    envVarName: string;
+  }[];
+  triggerFunctions?: string[];
```
same

iliapolo commented on packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/backend/synthesizer.ts:
```diff
@@ -826,18 +834,269 @@ export class BackendSynthesizer {
+
+      // Generate tables
+      renderArgs.storage.dynamoTables.forEach((table: DynamoDBTableDefinition) => {
+        const sanitizedTableName = this.sanitizeVariableName(table.tableName);
```
Remove the `-${env}` suffix from the table name before sanitizing. If we can avoid environment specifiers in the gen2 code we should. i.e instead of `tableName-main` we should just have `tableName`. 

It makes the gen2 code less specific and more portable since it doesn't "hint" at any specific environment names.

dgandhi62 replied:
Made the change

iliapolo commented on packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/backend/synthesizer.ts:
```diff
@@ -826,18 +834,269 @@ export class BackendSynthesizer {
+
+        // Add Lambda permissions
+        table.lambdaPermissions?.forEach((perm) => {
+          const grantCall = factory.createExpressionStatement(
```
same

iliapolo commented on packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/codegen-head/app_storage_definition_fetcher.ts:
```diff
@@ -58,12 +61,131 @@ export class AppStorageDefinitionFetcher {
   };
 
+  // Check the properties
+  private parseDynamoDBTable = async (
```
`parse` usually means operating on a string, not fetching values from an external source.  

```suggestion
  private fetchDynamoDBTable = async (
```

iliapolo reviewed (changes requested):
Please also update the README of app-2 now that storage is generated. Mention what step in the migration process will not work.

dgandhi62 committed:
feat: remove old dynamoDB property

dgandhi62 committed:
chore: remove unused import

dgandhi62 commented on packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/core/migration-pipeline.ts:
```diff
@@ -106,6 +107,26 @@ export interface Gen2RenderingOptions {
+ * @param content - Original handler file content
+ * @returns Transformed content with ES module syntax
+ */
+const transformHandlerToESModule = (content: string): string => {
```
This was part of the functions implementation I had attempted. Removing this

dgandhi62 committed:
chore: remove function ESModule error handling to maintain separation of concerns

dgandhi62 committed:
chore: change function name to make it fetchDynamodbTable from parseDynamodbTbable

dgandhi62 committed:
feat: remove lambda trigger and permissions support

dgandhi62 committed:
feat: remove dynamodb table naming

dgandhi62 committed:
feat: remove lambda logic from fetcher and tests

dgandhi62 committed:
chore: remove unused import

dgandhi62 committed:
feat: update tests to match new naming pattern

dgandhi62 committed:
fix: resolve test failures in template-generator.test.ts

- Import setup-jest to load custom Jest matcher toBeACloudFormationCommand
- Fix TypeScript compilation errors for custom matcher types
- Ensure snapshot tests pass with proper test setup

dgandhi62 committed:
chore: fixed variable name

dgandhi62 committed:
feat: only create DynamoDB table variable when GSIs exist

Conditionally create table variable declaration only when GSIs are present
to avoid eslint violations for unused variables.

dgandhi62 commented on packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/backend/synthesizer.ts:
```diff
@@ -826,18 +834,269 @@ export class BackendSynthesizer {
+          [],
+          factory.createVariableDeclarationList(
+            [
+              factory.createVariableDeclaration(
```
Resolved

dgandhi62 committed:
fix: update snapshot tests

dgandhi62 committed:
chore: merge local with upstream
